src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+                mir: &mut Mir<'tcx>,
+                drop_clean: BasicBlock) {
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let return_block = BasicBlock::new(mir.basic_blocks().len());
+    mir.basic_blocks_mut().push(BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: TerminatorKind::Return,
+        }),
+        is_cleanup: false,
+    });
+
+    let mut cases: Vec<_> = transform.bb_targets.iter().filter_map(|(&(r, u), &s)| {
+        u.map(|d| (s, d))
+    }).collect();
+
+    cases.insert(0, (0, drop_clean));
+
+    // The poisoned state 1 falls through to the default case which is just to return
+
+    let switch = TerminatorKind::SwitchInt {
+        discr: Operand::Consume(transform.make_field(transform.state_field, tcx.types.u32)),
+        switch_ty: tcx.types.u32,
+        values: Cow::from(cases.iter().map(|&(i, _)| {
+                ConstInt::U32(i)
+            }).collect::<Vec<_>>()),
+        targets: cases.iter().map(|&(_, d)| d).chain(once(return_block)).collect(),
+    };
+
+    insert_entry_point(mir, BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: switch,
+        }),
+        is_cleanup: false,
+    });
+
+    for block in mir.basic_blocks_mut() {
+        let kind = &mut block.terminator_mut().kind;
+        if let TerminatorKind::GeneratorDrop = *kind {
+            *kind = TerminatorKind::Return;
+        }
+    }
+
+    // Replace the return variable
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    mir.return_ty = tcx.mk_nil();
+    mir.local_decls[RETURN_POINTER] = LocalDecl {
+        mutability: Mutability::Mut,
+        ty: tcx.mk_nil(),
+        name: None,
+        source_info,
+        internal: false,
+        is_user_variable: false,
+    };
+
+    make_generator_state_argument_indirect(tcx, def_id, mir);
+
+    // Change the generator argument from &mut to *mut
+    mir.local_decls[Local::new(1)] = LocalDecl {
+        mutability: Mutability::Mut,
+        ty: tcx.mk_ptr(ty::TypeAndMut {
+            ty: gen_ty,
+            mutbl: hir::Mutability::MutMutable,
+        }),
+        name: None,
+        source_info,
+        internal: false,
+        is_user_variable: false,
+    };
+
+    // Make sure we remove dead blocks to remove
+    // unrelated code from the resume part of the function
+    simplify::remove_dead_blocks(mir);
+
+    dump_mir(tcx, None, "generator_drop", &0, source, mir);
+}
+
+fn insert_resume_after_return<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                        def_id: DefId,
+                                        mir: &mut Mir<'tcx>) -> Option<BasicBlock> {
+    let drop_arg = mir.local_decls.raw[2].ty.needs_drop(tcx, tcx.param_env(def_id));
```

Comment:
`2` is the generator argument, @alexcrichton kindly tried to rip out the support for this.